runtime/runtime.go

```diff
@@ -136,12 +137,27 @@ func (rm *RuntimeManager) FunctionRunWrapperWithHooks(fn interface{}) {
 			rm.FuncContext.WithError(err)
 
 		} else if rm.FuncContext.GetSyncRequest().Request != nil {
-
-			body, _ := ioutil.ReadAll(rm.FuncContext.GetSyncRequest().Request.Body)
+			var body []byte
+			// if it is a cloud event, we extra the cloudevent data as user data, and pass the raw cloud event in ctx
+			// if it is a http request, we pass the request body as user data and create a dummy cloud event with user data
+			r := rm.FuncContext.GetSyncRequest().Request
+			msg := cehttp.NewMessageFromHttpRequest(r)
```

Comment:
Does this mean we only support binary format of cloudevent, not structured? https://knative.tips/eventing/events/cloudevents/

Reply:
from my testing, both are supported. I will add one example in the e2e.

```sh

❯ curl -X POST -H "Content-type: application/json" -H "Accept: application/json" -s "http://localhost:8080/bar/openfunction" -H "Ce-Specversion: 1.0" -H "Ce-Type: io.openfunction.samples.helloworld" -H "Ce-Source: io.openfunction.samples/helloworldsource" -H "Ce-Id: 536808d3-88be-4077-9d7a-a3f162705f79" -H "Content-Type: application/json" -d 'hello'
{"hello":"openfunction"}%   

❯ curl -X POST "http://localhost:8080/bar/openfunction" \
   -H "Content-Type: application/cloudevents+json" \
   -d '{
        "specversion" : "1.0",
        "type" : "example.com.cloud.event",
        "source" : "https://example.com/cloudevents/pull",
        "subject" : "123",
        "id" : "A234-1234-1234",
        "time" : "2018-04-05T17:31:00Z",
        "data" : "hello"
}'
{" \"hello\"":"openfunction"}% 
```

Reply:
great！